services/src/main/java/org/keycloak/authentication/forms/RegistrationFormsHelper.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ *  and other contributors as indicated by the @author tags.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ */
+
+package org.keycloak.authentication.forms;
+
+import jakarta.ws.rs.core.MultivaluedHashMap;
+import jakarta.ws.rs.core.MultivaluedMap;
+import org.keycloak.authentication.FormContext;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.userprofile.UserProfile;
+import org.keycloak.userprofile.UserProfileContext;
+import org.keycloak.userprofile.UserProfileProvider;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class RegistrationFormsHelper {
+
+    /**
+     * Make sure that some fields (EG. password) are filtered before being send to user profile
+     */
+    private static MultivaluedMap<String, String> normalizeFormParameters(MultivaluedMap<String, String> formParams) {
+        MultivaluedHashMap<String, String> copy = new MultivaluedHashMap<>(formParams);
+
+        // Remove "password" and "password-confirm" to avoid leaking them in the user-profile data
+        copy.remove(RegistrationPage.FIELD_PASSWORD);
+        copy.remove(RegistrationPage.FIELD_PASSWORD_CONFIRM);
+
+        return copy;
+    }
+
+    /**
+     * Get user profile instance for current HTTP request (KeycloakSession) and for given context. This assumes that there is
+     * single user registered within HTTP request, which is always the case in Keycloak
+     */
+    public static UserProfile getOrCreateUserProfile(FormContext formContext, UserProfileContext upContext, MultivaluedMap<String, String> formData) {
```

Comment:
I understand this method is to allow sharing of the instance between both actions. But I'm wondering, is there any way to avoid using 3 different actions during registration but a single one?

Reply:
That is good point about different actions. I was wondering the same when looking at this as one of the options :-)

But at this PR, it is not about sharing the instance between different actions, but more about sharing between methods `validate()` and `success()` of the same action.

Currently there is:
- The `RegistrationUserCreation` uses UserProfileContext `REGISTRATION_USER_CREATION`
- The `RegistrationProfile` uses UserProfileContext `REGISTRATION_PROFILE`

Before my PR, there are 4 instances of UserProfile created during single registration at these places:
1) `RegistrationUserCreation.validate()`
2) `RegistrationProfile.validate()`
3) `RegistrationUserCreation.success()
4) `RegistrationProfile.success()`

In my PR, I've reduced it to 2 instances instead of 4 instances to avoid calling `validate()` multiple times, but still keep both actions and hence 2 separate UserProfile instances (one for REGISTRATION_USER_CREATION and one for REGISTRATION_PROFILE.

But yes, I can also think about removing one of `RegisterProfile` or `RegistrationUserCreation`, which will also allow to have only single context for registration like `REGISTRATION_USER_CREATION` instead of `REGISTRATION_PROFILE`. This seems to me clearer to merge both registration contexts and have only single context `UserProfileContext.REGISTER`. Also I think that it will simplify handling of userProfile during registration in general.

IMO there were 2 separate actions like `RegisterProfile` and `RegistrationUserCreation` needed before we have user-profile, but maybe now it's good time to remove one of them since we have user-profile? Should I proceed this way in this PR?

Reply:
Sounds great ... It simplify things a lot.

Reply:
Ok, will update PR in this regard